src/os/FileStore.cc

```diff
@@ -2883,76 +2883,167 @@ int FileStore::read(
+                                  bool *fallback_to_fiemap)
+{
+  off_t hole_pos, data_pos;
+  int r = 0;
 
-      dout(10) << "FileStore::fiemap() fm_mapped_extents=" << fiemap->fm_mapped_extents
-	       << " fe_logical=" << extent->fe_logical << " fe_length=" << extent->fe_length << dendl;
+  // If compiled on an OS with SEEK_HOLE/SEEK_DATA support, but running
+  // on an OS that doesn't support SEEK_HOLE/SEEK_DATA, EINVAL is returned.
+  // Fall back to use fiemap.
+  hole_pos = lseek(fd, 0, SEEK_HOLE);
+  if (hole_pos < 0) {
+    if (errno == EINVAL) {
+      *fallback_to_fiemap = true;
+    }
+    r = -errno;
+    return r;
+  }
 
-      /* try to merge extents */
-      while ((i < fiemap->fm_mapped_extents - 1) &&
-             (extent->fe_logical + extent->fe_length == next->fe_logical)) {
-          next->fe_length += extent->fe_length;
-          next->fe_logical = extent->fe_logical;
-          extent = next;
-          next = extent + 1;
-          i++;
+  // Seek back to pos 0 first if there is a hole.
+  if (hole_pos > 0) {
+    off_t tmp_pos;
+    tmp_pos = lseek(fd, 0, SEEK_SET);
+    if (tmp_pos != 0) {
+      r = -errno;
+      return r;
+    }
+  }
+
+  // If lseek fails with errno setting to be ENXIO, this means the current
+  // file offset is beyond the end of the file.
+  off_t start = offset;
+  while(start < (off_t)(offset + len)) {
+    data_pos = lseek(fd, start, SEEK_DATA);
+    if (data_pos < 0) {
+      if (errno == ENXIO)
+        break;
+      else {
+        r = -errno;
+	return r;
       }
+    } else if (data_pos > (off_t)(offset + len)) {
+      break;
+    }
 
-      if (extent->fe_logical + extent->fe_length > offset + len)
-        extent->fe_length = offset + len - extent->fe_logical;
-      exomap[extent->fe_logical] = extent->fe_length;
-      i++;
-      extent++;
+    hole_pos = lseek(fd, data_pos, SEEK_HOLE);
+    if (hole_pos < 0) {
+      if (errno == ENXIO) {
+        break;
+      } else {
+        r = -errno;
+	return r;
+      }
     }
-    free(fiemap);
+
+    if (hole_pos > (off_t)(offset + len)) {
+      (*m)[data_pos] = offset + len - data_pos;
+      break;
+    }
+    (*m)[data_pos] = hole_pos - data_pos;
+    start = hole_pos;
   }
 
+  return r;
+}
+#endif
+
+int FileStore::fiemap(coll_t cid, const ghobject_t& oid,
+                    uint64_t offset, size_t len,
+                    bufferlist& bl)
+{
+  tracepoint(objectstore, fiemap_enter, cid.c_str(), offset, len);
+
+  if (len <= (size_t)m_filestore_fiemap_threshold) {
+    map<uint64_t, uint64_t> m;
+    m[offset] = len;
+    ::encode(m, bl);
+    return 0;
+  }
+
+  dout(15) << "fiemap " << cid << "/" << oid << " " << offset << "~" << len << dendl;
+
+  map<uint64_t, uint64_t> exomap;
+  bool fallback_to_fiemap = false;
+  FDRef fd;
+
+  int r = lfn_open(cid, oid, false, &fd);
+  if (r < 0) {
+    dout(10) << "read couldn't open " << cid << "/" << oid << ": " << cpp_strerror(r) << dendl;
+    goto done;
+  }
+  
```

Comment:
I think we don't want to fallback to fiemap each time if doesn't support SEEK_DATA/SEEK_HOLE. We could detect this in "init" process.

And I think we need to add config value to make sure whether use SEEK_DATA/SEEK_HOLE or not just like FIEMAP, this logic and detect logic can be decided at the beginning